# alntools
utilities for processing .1aln files from [FastGA](https://github.com/thegenemyers/FASTGA), [FasTAN](https://github.com/thegenemyers/FASTAN) etc.

Make with `make` and install with `make install`.  If you want to install elsewhere than the default `~/bin` directory then `make DESTDIR=/your/preferred/location`.

Current contents are:

## tanbed

converts a .1aln file generated by [FasTAN](https://github.com/thegenemyers/FASTAN) into a [BED](https://en.wikipedia.org/wiki/BED_(file_format)) file.  Example usage is: 

```
> tanbed mGorGor-tan.1aln > mGorGor-tan.bed
read GDB from mGorGor-tan.1aln : 26 seqs 28 contigs (2 gaps), totSeq 3545850636 totCtg 3543850636 (99.944%)
processed 636143 alignments total length 708296862 from mGorGor-tan.1aln length 3545850636 (20.0 %)
```

## gdbmask

masks a .1gdb file as used by [FastGA -M](https://github.com/thegenemyers/FASTGA) with a bed file, as for example generated by **tanbed** from [FasTAN](https://github.com/thegenemyers/FASTAN) output.  Example usage is: 

```
> gdbmask mGorGor.1gdb mGorGor-tan.bed
read GDB from mGorGor.1gdb : 26 seqs 28 contigs (2 gaps), totSeq 3545850636 totCtg 3543850636 (99.944%)
read 636143 bed lines from mGorGor-tan.bed
wrote GDB to mGorGor.1gdb : 26 seqs 28 contigs (2 gaps), totSeq 3545850636 totCtg 3543850636 (99.944%), 636143 masks totMask 708296862 (20.0%)
!! rerun GIXmake mGorGor.1gdb to apply new mask
```

Note that, as requested in the final line of the output, you need to run GIXmake again from the [FastGA package](https://github.com/thegenemyers/FASTGA) in order for this masking to take effect in subsequent `FastGA` runs, and you also need to remember to set the `-M` option for "use soft Masks" when running `FastGA`.

By default `gdbmask` overwrites the given `.1gdb` file.  If you wish to keep that and write a new `.1gdb` file then you can use option `[-o newfile.1gdb]`, but for downstream tools to subsequently use the resulting `newfile.1gdb` you will need to create by copying (or linking) a corresponding hidden `.newname.bps` file that contains the 2-bit compressed sequence.

## taco

TAndem repeat COmpresses a .fa.gz file using coordinates from a `.1aln` file made with [FasTAN](https://github.com/thegenemyers/FASTAN), leaving just the first unit of each tandem repeat. The idea is to use the resulting compressed sequences in subsequent genome alignments, e.g. using [FastGA](https://github.com/thegenemyers/FASTGA). The reasoning behind this is that tandem repeats expand and contract very rapidly, so are likely not to be the same length between two otherwise alignable sequences - in fact because of this expansion/contraction and gene conversion it may well be that copies within a sequences are more closely related to each other than copies between sequences. Also, pragmatically aligners including FastGA can become very inefficient when aligning tandem repeats to each other, exploring and potentially re-exploring exponentially many similar alternatives.  Example usage is: 

```
> taco
Usage: taco [-o <outFileName>] <input.1aln> <seqFile>
  taco stands for 'TAndem COmpress' (cf hoco for 'HOmopolymer COmpress')
  input.1aln should be created by FasTAN and the names and lengths must match to seqFile
  default outFileName is <seqFile-stem>-taco.1seq;  user given outFileName can end in .fa or .fa.gz or .1seq
> taco mGorGor-tan.1aln mGorGor.fa.gz
read GDB from mGorGor-tan.1aln : 26 seqs 28 contigs (2 gaps), totSeq 3545850636 totCtg 3543850636 (99.944%)
> seqstat mGorGor.fa.gz
fasta file, 26 sequences >= 0, 3545850636 total, 136378870.62 average, 16412 min, 243847345 max
> seqstat mGorGor-taco.fa.gz
fasta file, 26 sequences >= 0, 2872001354 total, 110461590.54 average, 16412 min, 223911933 max
```

By default taco converts `XX.fa.gz` to `XX-taco.fa.gz`. Option `-o <name.fa.gz>` will write an alternative file name.

As of 16/11/2025, taco writes out a warning for each repeat in the input `.1aln` file that is contained within the first instance of another repeat, and does not compress it. There are not so many of these and they are typically short dinucleotide to tetranucleotide repeats. I aim to correctly compress these nested repeats and remove this warning. Also, there is not yet a tool to map the results of aligning taco sequences back into the original reference coordinates.

## svfind

retrieves large insertions and deletions from a .1aln file.  Example usage is:

```
> svfind -a ~/tmp/dmel-dmau-a.1sv -b ~/tmp/dmel-dmau-b.1sv dmel-dmau.1aln
read 130445 overlaps
user	0.100205	system	0.006640	elapsed 0.108578	alloc_max 6 MB	max_RSS	19562496
wrote 3355 insertions in ../Drosophila_melanogaster/Dmela-FKT.1gdb to /Users/rd109/tmp/dmel-dmau-a.1sv
user	0.696388	system	0.016602	elapsed 0.713271	alloc_max 107 MB	max_RSS	86114304
wrote 3438 insertions in ../Drosophila_mauritiana/Dmaur-FKT.1gdb to /Users/rd109/tmp/dmel-dmau-b.1sv
user	0.685140	system	0.014473	elapsed 0.699762	alloc_max 208 MB	max_RSS	71139328
Total resources used: user	1.481736	system	0.037810	elapsed 1.521707	alloc_max 208 MB	max_RSS	176816128
```
Note that the outputs are ONEcode files of type .1sv.  Run `ONEview -H outfile.1sv` for header information, and `ONEview -h outfile.1sv` for the results data.

There are lots of options.  Ask Pio Sierra <mailto:pas211@cam.ac.uk> for further information about them if you are interested in modifying them!

